Enable running bpfd as systemd service (#110)
redhat-et/bpfd: docs/admin/tutorial.md

# Tutorial

> 

Author: I have one. Maybe it would help to bold the 3 modes:
```suggestion
# Tutorial

This tutorial will show you how to use `bpfd`.
There are several ways to launch and interact with `bpfd` and `bpfctl`:

* **Basic Mode** - Run `bpfd` in foreground or background and straight from build directory.
  No `bpfd` or `bpfctl` users are created so always need `sudo` when executing `bptctl` commands.
  See [Basic Mode](#basic-mode).
* **`bpfctl` without `sudo`** - Run `bpfd` in foreground or background and straight from
  build directory, but also create the `bpfd` or `bpfctl` users and user groups.
  So if `bptctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.
  See [`bpfctl` without `sudo`](#bpfctl-without-sudo).
* **Systemd Service** - Run "bpfd" as a systemd service.
  See [Systemd Service](#systemd-service).

```

> 

Author: I'll also rename:
"Basic Mode" as "Priviledged Mode"
"`bpfctl` without `sudo`" as "Unpriviledged Mode"

This tutorial will show you how to use `bpfd`.
There are several ways to launch and interact with `bpfd` and `bpfctl`:

* Basic Mode - Run `bpfd` in foreground or background and straight from build directory.
  No `bpfd` or `bpfctl` users are created so always need `sudo` when executing `bptctl` commands.
  See [Basic Mode](#basic-mode).
* `bpfctl` without `sudo` - Run `bpfd` in foreground or background and straight from
  build directory, but also create the `bpfd` or `bpfctl` users and user groups.

>   So if `bptctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.

Reviewer: ```suggestion
  So if `bpfctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.
```

  See [`bpfctl` without `sudo`](#bpfctl-without-sudo).
* Systemd Service - Run "bpfd" as a systemd service.
  See [Systemd Service](#systemd-service).

## Basic Mode

### Step 0: Prerequisites

This tutorial uses examples from the [xdp-tutorial](https://github.com/xdp-project/xdp-tutorial).
You will need to check out the git repository and compile the examples.

### Step 1: Build `bpfd`

Perform the following steps to build `bpfd`.

```console
cd $HOME/src/bpfd/
cargo xtask build-ebpf --libbpf-dir $HOME/src/libbpf
cargo build
```

### Step 2: Setup `bpfd` environment

> `bpfd` uses mTLS for mutual authentication.

Reviewer: ```suggestion
`bpfd` uses mTLS for mutual authentication with clients.
```

In order to run and interact with `bpfd`, certificates must be created.

> Scripts to create self-sign certificates can be used in a development environment.

Reviewer: ```suggestion
Scripts to create a local certificate authority which is used to sign certs for bpfd and bpfctl can be used in a development environment.
```

I think is the right terminology, since "technically" they aren't self signed certs

Author: ack

(`/usr/lib/systemd/system/bpfd.service`).

### Step 0-1

Same as above.

### Step 2: Setup `bpfd` environment

To copy the `bpfd` and `bpfctl` binaries to `/usr/sbin/.` and set the user and user group for each,
and copy a default `bpfd.service` file to `/usr/lib/systemd/system/`, run the following command
(this command runs the same steps as the previous `setup.sh` commands plus some additional steps,
so run it in-place of  `setup.sh certs` or `setup.sh init`):

```console
sudo ./scripts/setup.sh install
```

Then add usergroup `bpfctl` to desired user if not already run and logout/login to apply:
```console
sudo usermod -a -G bpfctl \$USER
exit
<LOGIN>
```

To update the configuration settings associated with running `bpfd` as a service, edit the
service configuration file:

```console
sudo vi /usr/lib/systemd/system/bpfd.service
```

If `bpfd` or `bpfctl` is rebuilt, the following command can be run to install the update binaries
without tearing down the users and regenerating the certifications.
The `bpfd` service will is automatically restarted, so installed programs will need to be loaded again.

```console
sudo ./scripts/setup.sh reinstall
```

### Step 3: Start `bpfd`

> To manage `bpfd` as a systemd service, use `systemclt`:

Reviewer: ```suggestion
To manage `bpfd` as a systemd service, use `systemctl`:
```

Author: ack
